File: addVakta.go
package main

import (
	"encoding/json"
	"io/ioutil"
	"log"
	"os"
	"path/filepath"
	"sort"
	"strings"
)

type TextVariant struct {
	Ud    string `json:"ud"`
	Ur    string `json:"ur"`
	Ascii string `json:"ascii"`
}

type Entry struct {
	Text  TextVariant  `json:"text"`
	Vaktā *TextVariant `json:"vaktā,omitempty"`
}

func processJSONWithSpeaker(inputFolder, outputFolder string) error {
	// Ensure output folder exists
	if err := os.MkdirAll(outputFolder, os.ModePerm); err != nil {
		return err
	}

	// Read all JSON files in input folder
	files, err := ioutil.ReadDir(inputFolder)
	if err != nil {
		return err
	}

	for _, file := range files {
		if !strings.HasSuffix(file.Name(), ".json") {
			continue
		}

		// Full file paths
		inputPath := filepath.Join(inputFolder, file.Name())
		outputPath := filepath.Join(outputFolder, file.Name())

		// Read input file
		data, err := ioutil.ReadFile(inputPath)
		if err != nil {
			log.Printf("Error reading file %s: %v", file.Name(), err)
			continue
		}

		// Unmarshal JSON
		var jsonData map[string]Entry
		if err := json.Unmarshal(data, &jsonData); err != nil {
			log.Printf("Error unmarshaling JSON: %v", err)
			continue
		}

		// Sort keys
		var keys []string
		for k := range jsonData {
			keys = append(keys, k)
		}
		sort.Strings(keys)

		var lastSpeaker *TextVariant

		// Process keys
		for _, key := range keys {
			entry := jsonData[key]

			// If numeric key, update last speaker
			if isNumeric(key) {
				lastSpeaker = &entry.Text
			}

			// If alphabetic key and last speaker exists
			if !isNumeric(key) && lastSpeaker != nil {
				entry.Vaktā = lastSpeaker
				jsonData[key] = entry
			}
		}

		// Marshal processed data
		processedJSON, err := json.MarshalIndent(jsonData, "", "  ")
		if err != nil {
			log.Printf("Error marshaling JSON: %v", err)
			continue
		}

		// Write to output file
		if err := ioutil.WriteFile(outputPath, processedJSON, 0644); err != nil {
			log.Printf("Error writing file %s: %v", outputPath, err)
		}
	}

	return nil
}

func isNumeric(s string) bool {
	for _, c := range s {
		if c < '0' || c > '9' {
			return false
		}
	}
	return true
}

func main() {
	inputFolder := `D:\Projects\oper\decomposed`
	outputFolder := `D:\Projects\oper\vaktāsaha`

	if err := processJSONWithSpeaker(inputFolder, outputFolder); err != nil {
		log.Fatal(err)
	}
}

File: addVakta.py
import os
def process_json_with_speaker(input_json, output_file_path):
    processed_data = {}
    
    # Convert JSON to dictionary
    os.makedirs(output_folder, exist_ok=True)
    
    # Iterate through all JSON files in the input folder
    for filename in os.listdir(input_folder):
        if filename.endswith('.json'):
            input_file_path = os.path.join(input_folder, filename)
            output_file_path = os.path.join(output_folder, filename)
            
            # Read input JSON file
            with open(input_file_path, 'r', encoding='utf-8') as file:
                data = json.load(file)
    # print(input_json)
                data_dict = dict(input_json)
    
    # Track the last numeric key's text
                last_speaker = None
    
    # Sort keys to maintain order
                sorted_keys = sorted(data_dict.keys())
    
                for i, key in enumerate(sorted_keys):
        # If key is numeric, update last_speaker
                    if key.isnumeric():
                        last_speaker = data_dict[key]['text']
        
        # If key has alphabetic suffix and last_speaker exists
                    if not key.isnumeric() and last_speaker:

                        processed_data[key]= data[key]['vaktā'] = last_speaker
    
                        # **data[key],
                
                # Write processed data to output file
                    with open(output_file_path, 'w', encoding='utf-8') as outfile:
                        json.dump(processed_data, outfile, ensure_ascii=False, indent=2)
                        return processed_data

# Example usage
input_json = {
    "04001001": {
        "text": {
            "ud": "जनमेजय उवाच\n",
            "ur": "janamejaya uvāca\n",
            "ascii": "janamejaya uvAca\n"
        }
    },
    "04001001a": {
        "text": {
            "ud": "कथं विराटनगरे मम पूर्वपितामहाः\n",
            "ur": "kathaṁ virāṭanagare mama pūrvapitāmahāḥ\n",
            "ascii": "kathaM virATanagare mama pUrvapitAmahAH\n"
        }
    }
}
import json


input_folder = "D:\\Projects\\oper\\decomposed"
output_folder = "D:\\Projects\\oper\\vaktāsaha"
result = process_json_with_speaker(input_folder, output_folder)
print(json.dumps(result, ensure_ascii=False, indent=2))